### DISCUSSION TITLE
Is there a way to load .stl file with finer detail

### DISCUSSION THREAD

**Discussion by liruiw** [pyvista/pyvista#1453] [Q&A] [answered]
When initializing PolyData, it seems like there's some sampling on the original stl file. I wonder if there's a way to turn off the sampling, and get a very fine mesh of the original stl file as a numpy array

**Comment by akaszynski** [comment]
It should be resampling and there’s no option to turn it on or off.  Can you upload the file here for me to check it?

You can also check the number of faces and points in the polydata object and compare that with another mesh reading program.

**Comment by liruiw** [comment]
mesh = vtkInterface.PolyData(filename)
points = mesh.points
It's just these two lines of code. I've also been playing with vtk.STLReader and stl2numpy, they all provide the same number of points. In my case, I'd like to have a fine point cloud data in order to project those 3d points to image plane. If there's no option to adjust the resampling frequency, it seems like I have to repopulate the mesh myself. So, I wonder if the face and points returned by your program can be used to populate more points.

**Comment by akaszynski** [comment]
See the ``Subdivide`` function

```python
mesh = vtkInterface.PolyData(filename)
help(mesh.Subdivide)
```

**Comment by liruiw** [comment]
It's throwing the exception Exception: Subdivision filter must be one of the following: 'butterfly', 'loop', or 'linear'. And in vanilla vtk, it's saying "vtkLoopSubdivisionFilter (0x55f8a8aeb420): Dataset is non-manifold and cannot be subdivided. Edge shared by 3 cells." So subdivision seems not working.

**Comment by akaszynski** [comment]
VTK requires the surface to be manifold for subdivision.  Check out [pymeshfix](https://github.com/akaszynski/pymeshfix) if you'd like to make it manifold.

**Comment by liruiw** [comment]
Thank you, the pymeshfix package is working well. The problem is that after repairing, the manifold seems to contain only half of the size of the original model. In my case, before fixed it has around 7000 points, and after repairing, the top half (3500 points) is reserved. Is there a way to work around that?

**Comment by akaszynski** [comment]
Do you mean that the face order is reversed so that the face normals are pointed inward?  If so, you can correct the face ordering with:

```python
mesh.GenerateNormals(auto_orient_normals=True)

```

**Comment by liruiw** [comment]
Oh sorry. The face order looks fine to me. I mean the model size is reduced by half after repairing.

**Comment by akaszynski** [comment]
Are you sure those points weren’t dupicates?

**Comment by liruiw** [comment]
mesh = vtkInterface.PolyData(abs_name)		
meshfix = pymeshfix.MeshFix(mesh)
meshfix.Repair()
mesh = meshfix.mesh

Above is the code that I run. Those points were not duplicated (I print numberofpoints after first line, turns out to be 14224 points, and after the last line it becomes 3159). Do you know what's causing this? Also I wonder the meaning of the numbers printed when I used the meshfix. Thanks!

**Comment by akaszynski** [comment]
Can you zip the mesh and upload it?

**Comment by liruiw** [comment]
[mesh.zip](https://github.com/akaszynski/vtkInterface/files/2153007/mesh.zip)
Here it is. Thanks for your help.
Also I wonder if a z-buffer is implemented in python vtk or 3rd party libraries.

**Comment by akaszynski** [comment, marked as answer]
It looks like the input mesh contains several parts and isn't a single piece.  Additionally, the part contains several self-intersections that can't be repaired automatically with ``pymeshfix``.  I'd try outputting or generating the mesh as a single object.

Also, when debugging, you should use ``mesh.Plot()`` in order to visualize the mesh after repairing it to see what changed in each step.  ``pymeshfix`` removed all but one of the four isolated components, which is why only a quarter of the points remained after the "repair".

Z-buffer is implemented in VTK.
